chain/cosmos/tx.go

```diff
@@ -73,12 +71,17 @@ func (builder txBuilder) BuildTx(from, to address.Address, value, nonce, gasLimi
 
 	fees := Coins{Coin{
 		Denom:  builder.coinDenom,
-		Amount: pack.NewU64(gasPrice.Int().Uint64()),
+		Amount: pack.NewU64(gasPrice.Mul(gasLimit).Int().Uint64()),
 	}}
 
+	accountNumber, err := builder.client.AccountNumber(ctx, from)
```

Comment:
This is definitely not safe to do if this value can change over time. Under what conditions does the account number change?

Reply:
The account number does not change. But the reason why it was moved into the `BuildTx` method was that, an account number is assigned to an address only when an account is initialised for it. An account may be initialised only when a newly observed address receives funds from an existing account.

In our case, we know RenVM's key and even the deterministic address on Terra. But until RenVM receives funds from an existing account, we will never know what `account number` is assigned on-chain to RenVM's address.

Hence we cannot set our `account number` while instantiating a `Account.TxBuilder`. We have to do it only just before sending (and hence building) a transaction. Because at that moment, we know for sure that we must have received funds and so an account number has been assigned to us on-chain. This account number is a part of the ledger and does not change. It is a one-to-one mapping from address to account number that is network-specific.

So fetching the account number while building our transaction is safe.